core/src/main/java/com/dtolabs/rundeck/core/utils/WaitUtils.java

```diff
@@ -0,0 +1,38 @@
+package com.dtolabs.rundeck.core.utils;
+
+import java.time.Duration;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+public class WaitUtils {
+
+    /**
+     * Waits for the resource to be in the expected state by retrieving it and evaluating its acceptance as many times as needed.
+     *
+     * @param retryableResourceRetriever executed multiple times to retrieve the current state of the R resource.
+     * @param resourceAcceptanceEvaluator evaluates the state of the resource and returns true if the resource is accepted.
+     * @param timeout max duration to wait for the resource to reach the expected state.
+     * @param checkPeriod time to wait between each check.
+     * @return the resource that met the acceptance criteria.
+     * @throws ResourceAcceptanceTimeoutException if the timeout is reached waiting for the resource to reach the expected state.
+     * @throws InterruptedException if the thread sleep was interrupted.
+     */
+    public static <R> R waitFor(
```

Comment:
This does seem like a useful helper to have. There seem to be a number of places in the code-base that individually implement a retry with backoff.

We definitely don't have to do it in this implementation but a future improvement might be to add a growing backoff and a jitter around the `checkPeriod` duration.

Reply:
IMO, we should be using a proper retry library with all the bells and whistles instead of trying to implement our own. However, this is a starting point with minimal effort.

Reply:
I'm sure there's half a dozen other hand-rolled versions elsewhere in the code 😄